src/main/java/com/google/devtools/build/lib/remote/RemoteActionFileSystem.java

```diff
@@ -438,6 +449,47 @@ private void downloadFileIfRemote(PathFragment path) throws IOException {
     }
   }
 
+  @Override
+  public void renameTo(PathFragment sourcePath, PathFragment targetPath) throws IOException {
```

Comment:
Should we assert that `sourcePath` and `targetPath` are under `outputBase`? (I can't imagine a legitimate use case for moving to or from the source tree.) And in that case, do we ever need to fall back to the local filesystem?

Reply:
Added assertions for `sourcePath` and `targetPath`. Considering dynamic execution, both local and remote could have files so we need to try to rename on both. Also if the file is in the remote AND local (because it was staged), we need to keep them in sync.